### DISCUSSION TITLE
Idiomatic Implementation of `reactive_stores::PatchField` for non-primative types & enums?

### DISCUSSION THREAD

**Discussion by marcuswhybrow** [leptos-rs/leptos#3345] [Q&A] [answered]
Hi, I've got `reactive_stores::Store` working in my project to great effect. But I'm running into a (skill?) issue with getting `reactive_stores::Patch` implemented. Deriving `Patch` for a struct consisting entirely of "primitive" types will compile, but introducing a `std::option::Option` results in a compile time error. For example...

```rust
#[derive(Patch)]
struct Example {
    a: Option<String>,
}
```

produces compiler error: `` the trait `PatchField` is not implemented for `std::option::Option<std::string::String>` ``.

Digging into the `reactive_stores` source code I've established that, indeed, `PatchField` is only [implemented for specific primitive types](https://github.com/leptos-rs/leptos/blob/e1e90b859526a8b0f082496415dfe54f81583301/reactive_stores/src/patch.rs#L75-L114), and that doesn't include `Option`.

Rust (1.84.0-nightly) doesn't allow implementing external traits such as `PatchField` for generic types, so I can't do this:

```rust
impl<T> Patch for Option<T> { ... }

// nor this
impl Patch for Option<String> { ... } 
```

A common work around I've encountered to Rust's strictness r.e. external trait implementations is to wrap an external type in a new local type and implement the external trait for _it_. And indeed, this satisfies the compiler...

```rust
#[derive(Patch)]
struct Example {
    a: Patchable<Option<String>>,
}

#[derive(PartialEq)]
struct Patchable<T>(T);

impl<T> PatchField for Pachable<T> where T: PartialEq { ... }
```

Am I going about this the right way? My specific use-case requires `std::option::Option` and `std::path::PathBuf`. Perhaps the vision for a `Store` is to focus on using the correct primitive values to simplify the implementation. For example `PathBuf` could be replaced with `String`. `Option` on the other had has no lossless primative alternative.

Is this the idiomatic way to implement `PatchField` for generic types and enums, etc? 

Cheers,
Marcus

**Comment by gbj** [comment, marked as answer]
This is definitely just a case of me publishing a fairly experimental crate without implementing the trait for all the types it really should be implemented for, not a case of you running into a skill issue!

If you have a decent implementation of `PatchField` for `Option<_>` types, a PR would be welcome. If not, maybe just a "feature request" kind of issue so I or someone else can come along and implement it?

Purely an issue of time constraints on my part.
